### README for synonymous_substitution_rates

## Backtranslate_Orthogroup_TK.py
Tom Kono's python script to backtranslate amino acid sequences to nucleotide sequences

## convert_fasta_to_phylip.sh
Shell script to launch the python script _convert_fasta_to_phylip.py_

## convert_fasta_to_phylip.py
Python script to convert FASTA format to PHYLIP format

## create_yn00_control_files.py
Python script that automaticaly creates individual (hard-coded) control files for the yn00 program of PAML

## edit_names_in_phylip_file.py
Python script to edit the names of the PHYLIP file (changing gene name to species name)

## filter_fasta.py
Python script to filter FASTA file to retain only sequences we are interested in. (This one is outdated and replaced by _filter_fasta_zizania_duplications.py_).

## filter_fasta_zizania_duplications.py
Python script that is nearly identical to _filter_fasta.py_ but update to filter for _Zizania_-specific duplications

## filter_phylip_seqfiles.py
Python script written to further filter PHYLIP files, but did not end up using this script in my analysis.

## parse_paml_output.py
Python script to parse PAML output and write to a file for plotting later

## plot_synonymous_substitution_values.py
Python script to take parsed PAML output and plot it.

## run_backtranslate.sh
Shell script that runs the script _Backtranslate_Orthogroup_TK.py

## run_clustal_omega.sh
Shell script to run Clustal Omega (sequence alignment)

## run_create_yn00_control_files.sh
Shell script to launch the script _create_yn00_control_files.py_

## run_filter_fasta.sh
Shell script to launch the script _filter_fasta.py_

## run_filter_phylip_seqfiles.sh
Shell script to launch the script _filter_phylip_seqfiles.py_

## run_paml.sh
Shell script to run PAML (codeml)

## run_parse_paml_output.sh
Shell script to launch the script _parse_paml_output.py_

## run_yn00.sh
Shell script to launch the yn00 program of PAML

## setup_orthogroup_directories.sh
Shell script to set up orthogroup directory structure.
